llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -3045,9 +3205,28 @@ SDValue NVPTXTargetLowering::LowerFormalArguments(
                           DAG.getConstant(Offsets[VecIdx], dl, PtrVT));
           Value *srcValue = Constant::getNullValue(PointerType::get(
               EltVT.getTypeForEVT(F->getContext()), ADDRESS_SPACE_PARAM));
+
+          MaybeAlign PartAlign;
+          if (NumElts == 1) {
+            if (aggregateIsPacked) {
+              PartAlign = Align(1);
+            } else if (Offsets[parti] == 0) {
+              if (MaybeAlign ParamAlign = PAL.getParamAlignment(i)) {
+                PartAlign = ParamAlign.value();
+              } else {
+                PartAlign =
+                    DL.getABITypeAlign(EltVT.getTypeForEVT(F->getContext()));
+              }
+            } else {
+              PartAlign =
+                  DL.getABITypeAlign(EltVT.getTypeForEVT(F->getContext()));
+            }
+            PartAlign = commonAlignment(PartAlign.valueOrOne(), Offsets[parti]);
+          } else {
+            PartAlign = MaybeAlign(aggregateIsPacked);
+          }
```

Comment:
I think a lambda+early return may be a bit more readable.
```
PartsAlign = [](){
  if (NumElts != 1)
     return MaybeAlign(aggregateIsPacked);
  if (aggregateIsPacked)
     return Align(1);
  if (Offsets[parti] != 0)
     return DL.getABITypeAlign(EltVT.getTypeForEVT(F->getContext()));
  if (MaybeAlign ParamAlign = PAL.getParamAlignment(i))
      return ParamAlign.value();
    ...
  
}();
```

Reply:
I agree this is a bit ugly. Unfortunately, I think line 3224 means that we won't be able to significantly clean it up this way. 
```c++
 PartAlign = commonAlignment(PartAlign.valueOrOne(), Offsets[parti]);
```

Reply:
I'm missing something here. What makes this particular line special? 

Unless I've messed something up it all boils down to this:
```
    if (NumElts != 1) return MaybeAlign(aggregateIsPacked);
    if (aggregateIsPacked) return Align(1);
   MaybeAlign PartAlign =
        (Offsets[parti] == 0 && PAL.getParamAlignment(i))
            ? PAL.getParamAlignment(i).value()
            : DL.getABITypeAlign(EltVT.getTypeForEVT(F->getContext()));
    return commonAlignment(PartAlign.valueOrOne(), Offsets[parti]);
```

Reply:
What you have now looks right, I've updated to use that (with some small tweaks). I just meant that we wouldn't be able to do as much early exiting as you had in the first snippet.